plugins/woocommerce-blocks/assets/js/blocks/store-notices/frontend.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * External dependencies
+ */
+import { getContext, getElement, store } from '@woocommerce/interactivity';
+
+type Notice = {
+	notice: string;
+	type: 'error' | 'success' | 'notice';
+	dismissible: boolean;
+};
+
+type NoticeWithId = Notice & {
+	id: string;
+};
+
+type StoreNoticesState = {
+	notices: NoticeWithId[];
+};
+
+export type StoreNoticesStore = {
+	state: StoreNoticesState;
+	actions: {
+		addNotice: ( notice: Notice ) => string;
+		removeNotice: ( noticeId: string ) => void;
+	};
+	callbacks: {
+		dismissNotice: () => void;
+		isNoticeDismissible: () => boolean;
+		getNoticeClass: () => string;
+		getNoticeIconPath: () => string;
+		getNoticeRole: () => string;
+		renderNoticeContent: () => void;
+		scrollIntoView: () => void;
+	};
+};
+
+const ALERT_ICON_PATH =
+	'M12 3.2c-4.8 0-8.8 3.9-8.8 8.8 0 4.8 3.9 8.8 8.8 8.8 4.8 0 8.8-3.9 8.8-8.8 0-4.8-4-8.8-8.8-8.8zm0 16c-4 0-7.2-3.3-7.2-7.2C4.8 8 8 4.8 12 4.8s7.2 3.3 7.2 7.2c0 4-3.2 7.2-7.2 7.2zM11 17h2v-6h-2v6zm0-8h2V7h-2v2z';
+
+const ICON_PATHS = {
+	error: ALERT_ICON_PATH,
+	success: 'M16.7 7.1l-6.3 8.5-3.3-2.5-.9 1.2 4.5 3.4L17.9 8z',
+	notice: ALERT_ICON_PATH,
+};
+
+const generateNoticeId = () => {
+	// semi-random with low collision probability.
+	return `${ Date.now() }-${ Math.random()
+		.toString( 36 )
+		.substring( 2, 15 ) }`;
+};
+
+const { state } = store< StoreNoticesStore >( 'woocommerce/store-notices', {
+	state: {
+		notices: [],
+	},
+	actions: {
+		addNotice: ( notice: Notice ) => {
+			const noticeId = generateNoticeId();
+			state.notices = [
+				...state.notices,
+				{
+					...notice,
+					id: noticeId,
+				},
+			];
+
+			return noticeId;
```

Comment:
You don't need to use immutability with the Interactivity API. Not only is it unnecessary, but it also performs worse.	

```suggestion
			notice.id = generateNoticeId();
			state.notices.push( notice );
            return notice.id;
```

Reply:
Ok thanks, I didn't know that!  🗒️

Reply:
I've resolved this, but still using spread to assign id to notice variable mainly for type reasons, I don't want `Notice` to have `id?` because I don't want users to think they have an option to pass that.

Reply:
That's fine 🙂